libs/sync/src/lib.rs

```diff
@@ -0,0 +1,69 @@
+//! Only synchronisation primitive implementations should depend on this crate.
+//!
+//! If a crate uses a synchronisation primitive, it should depend on one of the
+//! following:
+//! - `sync_spin`
+//! - `sync_preemption`
+//! - `sync_irq`
+//! - `sync_block`
+
+#![no_std]
+
+pub mod mutex;
+
+pub use mutex::{Mutex, MutexGuard};
+
+/// A synchronisation flavour.
+pub trait Flavour {
+    /// Initial value for the lock data.
+    const INIT: Self::LockData;
+
+    /// Additional data stored on the lock.
+    type LockData;
+
+    type DeadlockPrevention: DeadlockPrevention;
+
+    /// Acquires the given mutex.
+    fn mutex_lock<'a, T>(
+        mutex: &'a mutex::SpinMutex<Self::DeadlockPrevention, T>,
+        data: &'a Self::LockData,
+    ) -> mutex::SpinMutexGuard<'a, Self::DeadlockPrevention, T>
+    where
+        Self: Sized;
+
+    /// Performs any necessary actions after unlocking the mutex.
+    fn post_unlock(mutex: &Self::LockData)
+    where
+        Self: Sized;
+}
+
+/// A deadlock prevention method.
+pub trait DeadlockPrevention {
+    /// Performs any necessary actions prior to locking.
+    fn enter();
+
+    /// Performs any necessary actions after unlocking.
+    fn exit();
```

Comment:
You've probably already considered this, but these trait methods are too limited because they have no stateful context. They need to offer a bit more complexity to be viable in other use cases, i.e., they need to be able to share state across the `enter` and `exit` calls. There are quite a few ways to do this, e.g., returning an object that is passed into exit.

Feel free to choose any method iface you want, but consider the most typical needs therein:
* maintaining a "guard" type (e.g., `HeldInterrupts`, `PreemptionGuard`) that is created in `enter()` and dropped in `exit()`
* performing an action that may require access to other mutex states (e.g., blocking a task)